# INF 8215 - Intelligence artif.: méthodes et algorithmes 
## Automne 2018 - TP1 - Méthodes de recherche 
### Membres de l'équipe
    - Fabrice Dugas (1690694)
    - Félix Boulet (1788287)
    - Yujia Ding (1801923)



## LE VÉLO À MONTRÉAL
Chaque année, Montréal accueille à peu près 10 millions de touristes. Soucieuse de la qualité de leur séjour, Tourisme Montréal a entamé un projet de développement d’une nouvelle application mobile afin d’assister les touristes lors de leurs déplacements dans la ville. Cette application a pour but d’aider l’utilisateur à planifier sa visite des importantes attractions de la ville, de la façon la plus efficace possible (ie, sur la durée la plus courte). Étant donné qu’il a été observé que le moyen de transport privilégié des touristes pour explorer Montréal est le vélo, cette application a pour but de générer des circuits cyclables de durée minimale. Plus précisément, étant donné une liste d’attractions munie de points de départ et d’arrivée, la tâche est de proposer, à chaque fois, un chemin qui passe par toutes les attractions indiquées une seule fois, qui débute au point de départ et qui s’achève au point d’arrivée et dont la durée de trajet est minimale.

<img src="images/montreal.png" alt="" width="800"/>

Le travail demandé dans ce TP est de développer l’algorithme interne de l’application. Nous explorerons trois mécanismes de résolution différents :
1. Définition et exploration naïve d’un arbre de recherche
2. Exploration plus efficace en utilisant l’algorithme A*
3. Optimisation locale en utilisant une métaheuristique de recherche à voisinage variable (Variable Neighborhood Search, VNS)

## PRÉSENTATION DU PROBLÈME
Une façon naturelle de représenter notre problème est d’utiliser un graphe $G=(V, A)$ dirigé et complet. Chaque sommet dans $V$ est une attraction donnée et chaque arc dans $A$ représente une piste cyclable entre deux attractions distinctes. Chaque paire de sommets $i$ et $j$ est reliée par une paire d’arcs $a_{ij}$ et $a_{ji}$ dont les poids respectifs $w(a_{ij})$ et $w(a_{ji})$ ne sont pas nécessairement égaux. Concrètement, ces poids représentent la durée du trajet d’un sommet à l’autre (ainsi, $w$ est telle que $w : A \to\mathbb R^+$).

La liste des attractions à visiter est indiquée comme la suite $P = (p_1, ..., p_m)$ où $p_1$ et $p_m$ sont les sommets de départ et d’arrivée, respectivement.

## 1. DÉFINITION ET EXPLORATION NAÏVE D’UN ARBRE DE RECHERCHE (5 points)
Définissons un arbre de recherche $\mathcal{T}$ où chaque nœud représente une solution partielle $S$. Soient $V(S) \subseteq V$ et $A(S) \subset A$ l’ensemble des sommets visités et l’ensemble des arêtes sélectionnées, respectivement. Ainsi, le coût d’une solution est donné par :
$$g(S) = \sum_{a \in A(S)} w(a)$$

Seule l’origine est visitée initialement. Ainsi, la racine de l’arbre de recherche contient une solution partielle vide $S_{\textrm{root}}$ telle que $V(S_{\textrm{root}})=\{p_1\}$ et $A(S_{\textrm{root}}) = \emptyset$.

<img src="images/tree1.png" alt="" width="100"/>

À la suite de cela, les nœuds subséquents dans l’arbre sont tous créés en ajoutant, à chaque solution partielle $S$, un sommet subséquent dans $P\backslash V(S)$ avec l’arc correspondant dans $A$ qui relie ce sommet à la dernière attraction visitée. Le sommet $p_m$ n’est ajouté qu’à la fin, lorsqu’il est le seul sommet non encore visité. Plus formellement, si on note le sommet à ajouter $c$ et le dernier sommet visité $c'$, alors la nouvelle solution partielle obtenue est $V(S) \gets V(S) \cup \{c\}$ et $A(S) \gets A(S) \cup \{(c’,c)\}$.

Ci-dessous est un exemple de l’arbre étendu depuis sa racine où $c'$ = $p_1$ :

<img src="images/tree2.png" alt="" width="400"/>

À la fin, les feuilles de l’arbre sont des solutions complètes :

<img src="images/tree3.png" alt="" width="600"/>

### 1.1 Code
La fonction fournie ci-dessous permet d’extraire d’un fichier un graphe qui répond aux spécifications détaillées plus haut. Cette fonction retourne une $\texttt{ndarray}$ ($\texttt{graph}$) de taille $|V|\times |V|$ où $\texttt{graph[i,j]}$ représente le temps nécessaire pour traverser la piste cyclable de $i$ vers $j$.



In [1]:
import numpy as np

def read_graph():
    return np.loadtxt("montreal", dtype='i', delimiter=',')

graph = read_graph()

Notre première tâche est de définir la classe qui représente une solution partielle. Son constructeur est donné et reçoit comme argument la liste des sommets (attractions $P$) à visiter et le graphe ($G$). Celui-ci crée la solution $S_{\textrm{root}}$ avec les attributs suivants :
- $\texttt{g}$ : le coût de la solution partielle
- $\texttt{visited}$ : représente $V(S)$, discuté plus haut. Par définition, $\mathtt{vistited[-1]}$ représente le dernier sommet ajouté, $ c $.
- $\texttt{not}\_\texttt{visited}$ : représente $P\backslash V(S)$
- $\texttt{graph}$: représente le graphe G

Ensuite, il est demandé d’implanter la méthode $\texttt{add}$ qui mets à jour la solution partielle en ajoutant une nouvelle attraction à visiter parmi la liste $\texttt{not}\_\texttt{visited}$. Cette méthode reçoit comme arguments l’index du sommet à visiter parmi $\texttt{not}\_\texttt{visited}$.

Implantez $\texttt{add}$ :

In [2]:
import copy

class Solution:
    def __init__(self, places, graph):
        """
        places: a list containing the indices of attractions to visit
        p1 = places[0]
        pm = places[-1]
        """
        self.g = 0 # current cost
        self.h = 0 # estimate
        self.graph = graph 
        self.visited = [places[0]] # list of already visited attractions
        self.not_visited = copy.deepcopy(places[1:]) # list of attractions not yet visited
        
    def add(self, idx):
        """
        Adds the point in position idx of not_visited list to the solution
        """
        ret = copy.deepcopy(self)
        end = ret.not_visited.pop(idx)
        begin = ret.visited[-1]
        cost = ret.graph[begin, end]
        ret.g += cost
        ret.visited.append(end)
        return ret
    
    def __𝚕𝚝__(𝚜𝚎𝚕𝚏, 𝚘𝚝𝚑𝚎𝚛):
        return (self.g + self.h) < (other.g + other.h)
    
    def swap(self, i, j):
        """
        Swaps two visited nodes at index i and j
        i and j cannot be the first or last element in visited
        """
        
        if i == 0 or i == len(self.visited)-1 or j == 0 or j == len(self.visited)-1:
            print("[WARN] SWAPPING BEGIN OR END NODES i: {}, j: {}".format(i,j))
        
        ret = copy.deepcopy(self)
        
        # actually swap the nodes
        tmp = ret.visited[i]
        ret.visited[i] = ret.visited[j]
        ret.visited[j] = tmp
        
        # calculate new cost
        # could be more efficient by adding to the current cost but YOLO
        self.g = 0
        for idx, place in enumerate(ret.visited):
            if idx == 0:
                continue
            first = ret.visited[idx-1]
            self.g += self.graph[first, place]
        
        return ret
        
        

La prochaine étape est d’implanter une stratégie de parcours de l’arbre de recherche. Une première méthode simple est naïve est de mettre en œuvre une recherche en largeur ([Breadth-first search](https://moodle.polymtl.ca/pluginfile.php/444662/mod_resource/content/1/recherche_en_largeur.mp4), BFS).

Implantez $\texttt{bfs}$ qui mets en œuvre cette recherche. Elle prend en arguments le graphe courant ainsi que la liste des attractions à visiter $P$ et elle retourne la meilleure solution trouvée.

In [3]:
from queue import Queue

def bfs(graph, places):
    """
    Returns the best solution which spans over all attractions indicated in 'places'
    """
    explored_node_count = 0
    best_cost = 100000000000
    frontier = Queue()
    solution = Solution(places, graph)
    best_solution = None
    frontier.put(solution)
    while frontier.qsize() > 0:
        solution = frontier.get()
        if len(solution.not_visited) == 0 and solution.g < best_cost:
            best_cost = solution.g
            best_solution = solution
        for idx, place in enumerate(solution.not_visited):
            if place != places[-1] or len(solution.not_visited) == 1:
                explored_node_count += 1
                frontier.put(solution.add(idx))
                
    return best_solution, explored_node_count

### 1.2 Expérimentations

On propose trois exemples d’illustration pour tester notre recherche en largeur. Le premier exemple prend en compte 7 attractions, le second 10 et le dernier 11. Vu que cette recherche énumère toutes les solutions possibles, le troisième exemple risque de prendre un temps considérable à s’achever.

Mettez en œuvre ces expériences et notez le nombre de nœuds explorés ainsi que le temps de calcul requis.

In [4]:
import time 

#test 1  --------------  OPT. SOL. = 27
start_time = time.time()
places=[0, 5, 13, 16, 6, 9, 4]
sol, explored = bfs(graph=graph, places=places)
print(sol.g)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

27
445
--- 0.03197526931762695 seconds ---


In [5]:
#test 2 -------------- OPT. SOL. = 30
start_time = time.time()
places=[0, 1, 4, 9, 20, 18, 16, 5, 13, 19]
sol, explored = bfs(graph=graph, places=places)
print(sol.g)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

30
149920
--- 4.935027122497559 seconds ---


In [6]:
#test 3 -------------- OPT. SOL. = 26
start_time = time.time()
places=[0, 2, 7, 13, 11, 16, 15, 7, 9, 8, 4]
sol, explored = bfs(graph=graph, places=places)
print(sol.g)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

26
1349289
--- 44.424275159835815 seconds ---


| Test | Nb. noeuds | Temps (s) |
|------|------------|-----------|
| 1    | 445        | 0.0150063 |
| 2    | 149920     | 4.0327137 |
| 3    | 1349289    | 39.531402 |

## 2. RECHERCHE GUIDÉE À L’AIDE DE L’ALGORITHME A\* (7.5 points)
Pour notre deuxième méthode de recherche, au lieu d’énumérer toutes les solutions possibles, nous effectuons une recherche guidée à l’aide de l’algorithme A\*. Comme vu en classe, A\* est une recherche où les nœuds à explorer sont priorisés en fonction du coût courant d’une solution $g(S)$ ainsi que d’une estimation du coût restant vers la solution finale donné par une heuristique $h(S)$.

Dans le cas d’une minimisation, $h(S)$ est une borne inférieure du coût réel restant et on priorise l’exploration des nœuds dont $f(S) = g(S)+h(S)$ est le plus petit. Avec cette méthode, la première solution complète trouvée est assurément la solution optimale.

Pour une solution donnée $S$ avec un dernier sommet visité $c$, une possible fonction $h$ est telle que :

$h(S) =$ Le poids du chemin le plus court entre $c$ et $p_m$ dans le sous graphe $G_S$ contenant les sommets $P\backslash V(S) \cup \{c\}$

Remarque que ce chemin le plus court utilisé dans le calcul de l’estimation $h$ entre l’attraction courante et l’arrivée ne passera pas nécessairement pas tous les sommets restants.


Notre algorithme A\* se présente comme ceci :
1. Définir l’arbre de recherche $\mathcal{T}$ exactement comme auparavant. Le calcul de $h$ pour la solution initiale est inutile : c’est la seule solution qu’on a.
2. Sélectionner le meilleur nœud candidat pour expansion. La solution partielle $S_b$ de ce nœud candidat est telle que :

   $$ f(S_b) \leq f(S) \quad \forall S \in \mathcal{T} \qquad S_B, S \text{ pas encore sélectionnés}$$

   Si $S_b$ est une solution complète, l’algorithme s’arrête et $S_b$ est assurément la solution optimale, sinon on continue à l’étape 3.
3. Créer des solutions subséquentes qui connectent la dernière attraction visitée à chacune des attractions restantes. Attention, on ignore l’arrivée tant que celle-ci n’est pas la seule qui reste.
 - Mettez à jour les listes des sommets visités et non visités
 - Calculez $g$ et $h$ pour chaque solution
 - Insérer la nouvelle solution partielle dans l’arbre.
4. Répéter 2 et 3.


### 2.1 Code
Commençons d’abord par compléter la classe $\texttt{Solution}$ pour prendre en compte les changements nécessaires à A\* (on a besoin notamment d’un attribut supplémentaire pour l’estimation $h$).

On verra plus tard que A\* s’implante à l’aide d’une file de priorité (priority queue). Pour que celle-ci marche, il est nécessaire de surcharger (overload) l’opérateur de comparaison « < » relatif à nos objets $\texttt{Solution}$. En sachant ce qui fait qu'une solution est meilleure qu’une autre pour l'exploration, implanter la méthode $\_\_\texttt{lt}\_\_$ dans $\texttt{Solution}$. Son prototype est $\_\_\texttt{lt}\_\_\texttt{(self, other)}$.

Maintenant, nous devons implanter la fonction d’estimation $h$. Pour cela, on utilise l’[algorithme de Dijkstra](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm) pour trouver le chemin le plus court entre la dernière attraction visitée $c$ et l’arrivée $p_m$. Il est possible d’adapter cet algorithme pour qu’il s’arrête dès que le chemin le plus court entre c et pm est trouvé.

**Prescriptions d’implantation :**
- Appliquer Dijkstra pour trouver le chemin le plus court entre $c$ et $p_m$
- Retourner le poids de ce chemin

In [7]:
import heapq

def fastest_path_estimation(sol):
    """
    Returns the time spent on the fastest path between 
    the current vertex c and the ending vertex pm
    """
    if not sol.not_visited:
        return 0
    
    c = sol.visited[-1]
    pm = sol.not_visited[-1]
    
    # Array of tuples of the form (cost, node)
    q = [(0, c)]
    
    # Map the array to a set of tuples of the form (None, x) where x is the node id.
    remaining = set(map(lambda x: (None, x), sol.not_visited))
    processed = set()
    
    # While there are nodes in the priority queue
    while q:
        (cost, node) = heapq.heappop(q) # Pop the first one
        if node not in processed: # If the node wasn't processed already (at a better cost/priority)
            if node == pm: # If the node is the target, we're done
                return cost
            processed.add(node) # If not, add it to the processed nodes
            if node in remaining:
                remaining.remove(node) # Should always be true, but just in case
            for previous_cost, neighbor in remaining: # Check all remaining neighbors that have no shortest path yet
                cost_now  = cost + sol.graph[node, neighbor]
                if previous_cost is None or cost_now < previous_cost: # If it has a better cost, change it and push to the queue
                    previous_cost = cost_now
                    heapq.heappush(q, (cost_now, neighbor))
    return None # path is unreachable
    

Finalement, il est temps d’implanter A\*. On aura besoin d’une file de priorité qui retournera toujours le meilleur nœud candidat de $\mathcal{T}$ pour l’étendre (l’opérateur surchargé de comparaison assure cela).

**Prescriptions d’implantation (cf. détail des étapes de l’algorithme plus haut) :**
- Tant que les solutions extraites de la file de priorité ne sont pas complètes :
  *	Sélectionner et étendre le nœud extrait de la file comme détaillé plus haut
  * Calculer $g$ et $h$ pour chaque nouvelle solution partielle obtenue
  * Remettre ces solutions dans la file
- Retourner la première solution complète extraite de la file (c’est la solution optimale)

In [18]:
import heapq

def A_star(graph, places):
    """
    Performs the A* algorithm
    """
    
    explored_count = 0

    # blank solution
    root = Solution(graph=graph, places=places)

    # search tree T
    T = []
    heapq.heapify(T)
    heapq.heappush(T, root)

    # While there are nodes in the priority queue
    while T:
        solution = heapq.heappop(T) # Pop the first one
        if len(solution.not_visited) == 0:
            return solution, explored_count
        for idx, place in enumerate(solution.not_visited):
            if place != places[-1] or len(solution.not_visited) == 1:
                explored_count += 1
                temp_sol = solution.add(idx)
                temp_sol.h = fastest_path_estimation(temp_sol)
                #temp_sol.h = minimum_spanning_arborescence(temp_sol)
                heapq.heappush(T, temp_sol)
    return None, explored_count
    

### 2.2 Expérimentations

On ajoute un Quatrième exemple d’exécution avec 15 attractions. Là encore, mettez en œuvre ces expériences avec le nouvel algorithme A\* conçu et notez le nombre de nœuds explorés ainsi que le temps de calcul requis.

In [19]:
import time

#test 1  --------------  OPT. SOL. = 27
start_time = time.time()
places=[0, 5, 13, 16, 6, 9, 4]
astar_sol, explored = A_star(graph=graph, places=places)
print(astar_sol.g)
print(astar_sol.visited)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

27
[0, 5, 13, 16, 6, 9, 4]
42
--- 0.00447392463684082 seconds ---


In [20]:
#test 2  --------------  OPT. SOL. = 30
start_time = time.time()
places=[0, 1, 4, 9, 20, 18, 16, 5, 13, 19]
astar_sol, explored = A_star(graph=graph, places=places)
print(astar_sol.g)
print(astar_sol.visited)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

30
[0, 1, 4, 5, 9, 13, 16, 18, 20, 19]
75
--- 0.014627933502197266 seconds ---


In [21]:
#test 3  --------------  OPT. SOL. = 26
start_time = time.time()
places=[0, 2, 7, 13, 11, 16, 15, 7, 9, 8, 4]
astar_sol, explored = A_star(graph=graph, places=places)
print(astar_sol.g)
print(astar_sol.visited)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

26
[0, 2, 7, 7, 9, 13, 15, 16, 11, 8, 4]
150
--- 0.026469945907592773 seconds ---


In [22]:
#test 4  --------------  OPT. SOL. = 40
start_time = time.time()
places=[0, 2, 20, 3, 18, 12, 13, 5, 11, 16, 15, 4, 9, 14, 1]
astar_sol, explored = A_star(graph=graph, places=places)
print(astar_sol.g)
print(astar_sol.visited)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

40
[0, 3, 9, 13, 15, 18, 20, 16, 11, 12, 14, 5, 4, 2, 1]
722
--- 0.2680339813232422 seconds ---


#### Heuristique Dijkstra

| Test | Nb. noeuds | Temps (s) |
|------|------------|-----------|
| 1    | 120        | 0.0049860 |
| 2    | 1302       | 0.0540085 |
| 3    | 3831       | 0.1389732 |
| 4    | 624491     | 38.987563 |

#### Heuristique Minimum Spanning Arborescence

| Test | Nb. noeuds | Temps (s) |
|------|------------|-----------|
| 1    | 42         | 0.0029986 |
| 2    | 75         | 0.0109856 |
| 3    | 150        | 0.0199816 |
| 4    | 722        | 0.2069330 |



### 2.3 Une meilleure borne inférieure

Notre algorithme A\* est déjà beaucoup plus efficace qu’une recherche naïve. Cependant, la qualité de l’heuristique $h$ a un très grand impact sur la vitesse de A\*. Une heuristique plus serrée devrait accélérer A\* de façon significative. Notre estimation $h$ basée sur Dijkstra est très large à cause du fait qu’elle ne considère pas toutes les attractions restantes.

Une meilleure heuristique pourrait être basée sur la **Spanning Arborescence of Minimum Weight** qui s’apparente à une Minimum Spanning Tree pour graphes orientés. On propose de construire une telle Spanning Arborescence sur le reste des attractions $P\backslash V(S) \cup \{c\}$. Ici la racine est la dernière attraction visitée $c$. Une façon classique de résoudre ce problème est d’utiliser l’[algorithme de Edmonds](https://en.wikipedia.org/wiki/Edmonds%27_algorithm).

Implantez cet algorithme et refaites les expériences avec A\* en utilisant cette nouvelle heuristique :

In [23]:
from collections import namedtuple

Edge = namedtuple('Edge', 'u v w')

def edmonds(edges, root):
    minimum_edges = {}
    for edge in edges:
        if minimum_edges.get(edge.v) is None or edge.w < minimum_edges[edge.v].w:
            minimum_edges[edge.v] = edge
    minimum_edges[root] = Edge(-1, root, 0)

    groups = {}
    is_cycle_group = {}
    cycles = []
    stack = []
    index = {}
    lowlink = {}
    for edge in minimum_edges:
        if edge not in index and minimum_edges[edge].w > 0:
            find_cycle(edge, minimum_edges, cycles, stack, index, lowlink)
    
    groups[root] = root
    is_cycle_group[root] = False
    for edge in edges:
        if not groups.get(edge.v):
            groups[edge.v] = edge.v
            is_cycle_group[edge.v] = False
    
    for cycle in cycles:
        cycle_group = cycle[0].v
        for edge in cycle:
            if len(cycle) > 1:
                groups[edge.v] = cycle_group
                is_cycle_group[groups[edge.v]] = True

    if True not in is_cycle_group.values():
        result = 0
        for edge in minimum_edges:
            result += minimum_edges[edge].w
        return result
    
    result = 0
    for edge in minimum_edges:
        if (is_cycle_group[groups[minimum_edges[edge].v]]):
            result += minimum_edges[edge].w
            
    new_edges = []
    
    for edge in edges:
        found = False
        u = groups[edge.u]
        v = groups[edge.v]
        if u == v:
            continue
        elif is_cycle_group[v]:
            new_edges += [Edge(u, v, edge.w - minimum_edges[edge.v].w)]
        else:
            new_edges += [Edge(u, v, edge.w)]
    
    return result + edmonds(new_edges, groups[root])

        
# Tarjan's algorithm: https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
def find_cycle(node, edges, cycles, stack, index, lowlink):
    if node == -1:
        return
    index[node] = len(index) # The index keeps the order of discovery of the nodes
    lowlink[node] = index[node] # The lowlink keeps the smallest index that's reachable from our node
    stack += [node]
    neighbor = edges[node].u
    if neighbor not in index and neighbor != -1: # If the neighbor has not been visited yet,
        find_cycle(neighbor, edges, cycles, stack, index, lowlink) # Recursion! Let's go farther in our exploration (DFS)
        lowlink[node] = min(lowlink[neighbor], lowlink[node])
        # Take the minimum between the smallest index reachable from the neighbor and what we had before (transitivity)
    elif neighbor in stack: # If the neighbor is still in our DFS stack, meaning it could be 'better'
        lowlink[node] = min(lowlink[node], index[neighbor])
        # Take the minimum between the index of the neighbor and what we had before

    if lowlink[node] == index[node]: # If the smallest index reachable from the node is itself, we have a cycle!
        scc = [] # Strongly-connected component
        neighbor = None
        while node != neighbor and neighbor != -1: # As long as we don't find a neighbor that's the same as the node or the root,
            neighbor = stack.pop() # Pop what's in the cycle
            scc += [edges[neighbor]] # Add it to the strongly-connected component
        cycles += [scc] # Hey, we're done with this. Add it to the cycles.
        
def minimum_spanning_arborescence(sol):
    root = sol.visited[-1]
    minimum_edges = {}
    edges = []
    
    for node in sol.not_visited:
        for neighbor in sol.not_visited:
            if node != neighbor:
                edges += [Edge(node, neighbor, graph[node, neighbor])]
        edges += [Edge(root, node, graph[root, node])]
    
    
    return edmonds(edges, root)


## 3. RECHERCHE LOCALE À VOISINAGE VARIABLE  (7.5 points)

Cette fois-ci, au lieu de construire une solution optimale depuis une solution vide, on commence d’une solution complète, non-optimale, qu’on améliore à l’aide d’une recherche locale en utilisant une recherche locale à voisinage variable ([Variable Neighborhood Search](https://en.wikipedia.org/wiki/Variable_neighborhood_search), VNS).

<img src="images/vns.png" alt="" width="800"/>

### 3.1 Code

On commence par créer une solution initiale. Celle-ci est une suite ordonnée des attractions de $p_1$ à $p_m$ dans $P$. Pour cela, on fait appel à une [recherche en profondeur (Depth-First Search, DFS)](https://moodle.polymtl.ca/pluginfile.php/445484/mod_resource/content/1/recherche_en_profondeur.mp4) qu’on arrête aussitôt qu’une solution complète est trouvée. Pour aider à diversifier la recherche, la méthode permettant de générer une solution initiale peut être randomisée de telle sorte que l'algorithme VNS puisse lancer la recherche dans différentes régions de l'espace solution. Ainsi, dans la fonction DFS, la sélection de l'enfant pour continuer la recherche doit être aléatoire.

**Prescriptions d’implantation :**
- Mettre en œuvre une recherche en profondeur
- Créer un objet $\texttt{Solution}$ relatif à cette solution
- Ajuster les attributs de cet objet avec les bonnes valeurs de coûts et d’attractions visitées
- Retourner la solution trouvée.

In [24]:
from random import shuffle, randint

def initial_sol(graph, places):
    """
    Return a completed initial solution
    """
    return dfs(graph, places)
    

def dfs(graph, places):
    """
    Performs a Depth-First Search
    """    
    solution = Solution(places, graph)
    while len(solution.not_visited) > 0:
        not_visited = solution.not_visited
        if len(solution.not_visited) != 1:
            # on enleve la destination des choix pour la prochaine solution a visiter
            not_visited = solution.not_visited[0:-1]
            
        index = randint(0, len(not_visited)-1)
        solution = solution.add(index)
        
    return solution
    

Pour définir une VNS, il faut définir les $k_\textrm{max}$ voisinages de recherche locale possibles. Pour notre problème, une bonne et simple répartition des voisinages est telle qu’un voisinage $k$ correspond à la permutation de $k$-paires de sommets dans $V(S)$.

On appelle **shaking** l’étape de génération d’une solution dans le voisinage $k$. Le travail qui suit correspond à l’implantation de cette étape. $\texttt{shaking}$ admet 3 arguments que sont la solution de départ, l’indice du voisinage $k$ ainsi que le graph courant.

Attention, avant d’implanter $\texttt{shaking}$, il est nécessaire de créer une méthode $\texttt{swap}$ dans la classe $\texttt{Solution}$. Cette méthode permet de mettre en œuvre la permutation dans une solution donnée (en mettant à jour tous les attributs nécessaires pour que la solution soit cohérente).

**Prescriptions d’implantation de shaking :**
- Sélectionner au hasard deux indices $i$ et $j$ différents et tels que $i, j \in \{2,...,m-1\}$
- Faire une copie de la solution courante et faire la permutation
- Retourner la solution créée

In [25]:
def shaking(sol, k):
    """
    Returns a solution on the k-th neighborhood of sol
    """
    solution = copy.deepcopy(sol)
    for _ in range(k):
        i = randint(1,len(solution.visited)-2)
        j = randint(1,len(solution.visited)-2)
        while i == j:
            # assure que i sera different de j
            j = randint(1,len(solution.visited)-2)
        
        solution = solution.swap(i,j)
    return solution

Une dernière étape essentielle dans une VNS est l’application d’un algorithme de recherche locale à la solution issue du shaking. Pour cela, on propose la recherche locale 2-opt. Celle-ci intervertit deux arcs dans la solution, à la recherche d’une qui est meilleure.

Pour un sommet $ i $, soit $ i '$ le successeur immédiat de $ i $ dans la séquence $ V (S) $. L'algorithme 2-opt fonctionne comme suit: pour chaque paire de sommets non consécutifs $ i, j $, vérifiez si en échangeant la position des sommets $ i '$ et $ j $ entraîne une amélioration du coût de la solution. Si oui, effectuez cet échange. Ce processus se répète jusqu'à ce qu'il n'y ait plus d'échanges rentables. On réalise cette opération pour toutes les paires d’arcs éligibles à la recherche du plus petit coût.

<img src="images/2opt.png" alt="" width="800"/>

<img src="images/2opt2.png" alt="" width="800"/>


Implantez $\texttt{local}\_\texttt{search}\_\texttt{2opt}$. 

**Prescriptions d’implantation :**
- Considérer chaque paire d’indices $i = \{2,..,m-3\}$ and $j = \{i+2, m-1\}$
- Si l’échange donne un plus bas coût, on le réalise
- Répéter jusqu’à optimum local.

In [26]:
def local_search_2opt(sol):
    """
    Apply 2-opt local search over sol
    """
    best_solution = copy.deepcopy(sol)
    found_better_solution = True
    while found_better_solution:
        found_better_solution = False
        solution = copy.deepcopy(best_solution)
        # la prescriptions d'implantation est 1-indexed donc on doit utiliser le mm range -1
        for i in range(1,len(solution.visited)-4):
            for j in range(i+2,len(solution.visited)-2):
                i_prime = i+1
                neighbour = solution.swap(i_prime,j)
                if neighbour.g < best_solution.g:
                    best_solution = neighbour
                    found_better_solution = True
    
    return best_solution


Finalement, il est temps d'implanter notre VNS. La méthode $\texttt{vns}$ reçoit une solution complète, le graphe courant, le nombre maximal de voisinages et un temps de calcul limite. Celle-ci retourne la solution optimale trouvée

**Prescriptions d’implantation :**
- À chaque itération, la VNS génère une solution dans le k-ème voisinage (shaking) à partir de la meilleure solution courante et applique une recherche locale 2-opt dessus
- Si la nouvelle solution trouvée a un meilleur coût, mettre à jour la meilleure solution courante
- Répéter le processus jusqu'à $\texttt{t}\_\texttt{max}$

In [27]:
def vns(sol, k_max, t_max):
    """
    Performs the VNS algorithm
    """
    k = 1
    best_solution = copy.deepcopy(sol) # meilleure solution a vie
    start_time = time.time()
    end_time = start_time + t_max
    while time.time() < end_time:
        solution = shaking(best_solution, k)
        solution = local_search_2opt(solution)
        
        if solution.g < best_solution.g:
            best_solution = copy.deepcopy(solution)
            #print("VNS meilleure solution: {}".format(best_solution.g))
    
        # go to next neighborhood
        k = k % k_max + 1
    
    return best_solution

### 3.2 Experiments

Mettez en oeuvre la VNS sur les exemples d'illustration suivants et raportez les solutions obtenue:

In [32]:
# test 1  --------------  OPT. SOL. = 27
places=[0, 5, 13, 16, 6, 9, 4]
sol = initial_sol(graph=graph, places=places)
start_time = time.time()
vns_sol = vns(sol=sol, k_max=10, t_max=1)
print(vns_sol.g)
print(vns_sol.visited)
print("--- %s seconds ---" % (time.time() - start_time))

27
[0, 5, 6, 13, 16, 9, 4]
--- 1.0004453659057617 seconds ---


In [35]:
#test 2  --------------  OPT. SOL. = 30
places=[0, 1, 4, 9, 20, 18, 16, 5, 13, 19]
sol = initial_sol(graph=graph, places=places)

start_time = time.time()
vns_sol = vns(sol=sol, k_max=10, t_max=1)
print(vns_sol.g)
print(vns_sol.visited)

print("--- %s seconds ---" % (time.time() - start_time))

31
[0, 1, 4, 5, 16, 9, 13, 20, 18, 19]
--- 1.0004446506500244 seconds ---


In [41]:
# test 3  --------------  OPT. SOL. = 26
places=[0, 2, 7, 13, 11, 16, 15, 7, 9, 8, 4]
sol = initial_sol(graph=graph, places=places)

start_time = time.time()
vns_sol = vns(sol=sol, k_max=10, t_max=1)
print(vns_sol.g)
print(vns_sol.visited)
print("--- %s seconds ---" % (time.time() - start_time))

27
[0, 2, 7, 7, 8, 15, 11, 13, 16, 9, 4]
--- 1.0009167194366455 seconds ---


In [39]:
# test 4  --------------  OPT. SOL. = 40
places=[0, 2, 20, 3, 18, 12, 13, 5, 11, 16, 15, 4, 9, 14, 1]
sol = initial_sol(graph=graph, places=places)

start_time = time.time()
vns_sol = vns(sol=sol, k_max=10, t_max=1)
print(vns_sol.g)
print(vns_sol.visited)
print("--- %s seconds ---" % (time.time() - start_time))

48
[0, 3, 11, 12, 13, 4, 20, 15, 18, 14, 9, 16, 5, 2, 1]
--- 1.0008182525634766 seconds ---


#### Résultats

| Test | Noeuds visités | Coût |
|------|----------------|-----------|
| 1    | [0, 5, 6, 13, 16, 9, 4] | 27 |
| 2    | [0, 1, 4, 5, 16, 9, 13, 20, 18, 19] | 31 |
| 3    | [0, 2, 7, 7, 9, 11, 16, 15, 13, 8, 4] | 27 |
| 4    | [0, 3, 11, 12, 13, 4, 20, 15, 18, 14, 9, 16, 5, 2, 1] | 48 |


## 4. BONUS (1 point)

Expliquez dans quelle situation chacun des algorithmes développés est plus approprié (prenez en compte l’évolutivité du problème)

Pour le premier algorithme à implémenter, on explore toutes les nodes du graphe, ce qui fait que le temps d'exécution peut être très long si on a un graphe avec de nombreux sommets. On l'utilise dans des cas où il est crucial d'avoir la solution optimale et qu'on a plus de temps pour exécuter le programme. 

Par contre, le deuxième algorithme utilise des heuristiques pour élaguer l'arbre de recherche. Pour la même raison que le premier algorithme, on utilise cet algorithme pour assurer une solution optimale mais on est garantis d'avoir un temps d'exécution plus rapide que le premier. Il y avait quand même une différence observable entre les deux heuristiques étudiées pour le deuxième algorithme. En effet, le minimum spanning tree permettait à l'algorithme de ne visiter que quelques centaines d'attractions, alors que Dijkstra pouvait en faire visiter quelques centaines de milliers. Toutefois, le temps d'exécution peut quand même être long même avec l'heuristique la plus rapide lorsqu'il y a beaucoup d'attractions à visiter. De plus, on n'obtient pas de solution avant la fin de l'exécution de l'algorithme.

En dernier lieu, l'algorithme de recherche assure l'obtention d'une solution suboptimale rapidement. On peut également contrôler le temps d'exécution de ce dernier afin d'obtenir de meilleurs résultats si nécessaire. Cependant, il n'y a aucune garantie quant à l'optimalité de la solution obtenue. Dans le cas où on a beaucoup d'attractions à visiter, que peu de temps à consacrer à l'exécution de l'algorithme et que la solution optimale n'est pas nécessaire, il serait avantageux d'utiliser cette technique.

## 5 Date de remise
Tout devra être remis avant le 6 Octobre à 23h55. Tout travail en retard sera pénalisé d'une valeur de 10% par jour de retard.